contracts/subdomainregistrar/BaseSubdomainRegistrar.sol

```diff
@@ -0,0 +1,263 @@
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {ISubdomainPricer} from "./pricers/ISubdomainPricer.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+error Unavailable();
+error Unauthorised(bytes32 node);
+error NameNotRegistered();
+error InvalidTokenAddress(address);
+error NameNotSetup(bytes32 node);
+error DataMissing();
+error ParentExpired(bytes32 node);
+error ParentNotWrapped(bytes32 node);
+error DurationTooLong(bytes32 node);
+error ParentNameNotSetup(bytes32 parentNode);
+
+struct Name {
+    ISubdomainPricer pricer;
+    address beneficiary;
+    bool active;
+}
+
+abstract contract BaseSubdomainRegistrar {
+    mapping(bytes32 => Name) public names;
+    INameWrapper public immutable wrapper;
+    using Address for address;
+
+    event NameRegistered(bytes32 node, uint256 expiry);
+    event NameRenewed(bytes32 node, uint256 expiry);
+    event NameSetup(
+        bytes32 node,
+        address pricer,
+        address beneficiary,
+        bool active
+    );
+
+    uint64 internal GRACE_PERIOD = 90 days;
+
+    constructor(address _wrapper) {
+        wrapper = INameWrapper(_wrapper);
+    }
+
+    modifier authorised(bytes32 node) {
+        if (!wrapper.canModifyName(node, msg.sender)) {
+            revert Unauthorised(node);
+        }
+        _;
+    }
+    modifier canBeRegistered(bytes32 parentNode, uint64 duration) {
+        _checkParent(parentNode, duration);
+        _;
+    }
+
+    function available(bytes32 node) public view virtual returns (bool) {
+        try wrapper.getData(uint256(node)) returns (
+            address,
+            uint32,
+            uint64 expiry
+        ) {
+            return expiry < block.timestamp;
+        } catch {
+            return true;
+        }
+    }
+
+    function _setupDomain(
+        bytes32 node,
+        ISubdomainPricer pricer,
+        address beneficiary,
+        bool active
+    ) internal virtual authorised(node) {
+        names[node] = Name({
+            pricer: pricer,
+            beneficiary: beneficiary,
+            active: active
+        });
+        emit NameSetup(node, address(pricer), beneficiary, active);
+    }
+
+    function _batchRegister(
+        bytes32 parentNode,
+        string[] calldata labels,
+        address[] calldata addresses,
+        address resolver,
+        uint16 fuses,
+        uint64 duration,
+        bytes[][] calldata records
+    ) internal {
+        if (
+            labels.length != addresses.length || labels.length != records.length
+        ) {
+            revert DataMissing();
+        }
+
+        if (!names[parentNode].active) {
+            revert ParentNameNotSetup(parentNode);
+        }
+
+        _checkParent(parentNode, duration);
+
+        _batchPayBeneficiary(parentNode, labels, duration);
+
+        //double loop to prevent re-entrancy because _register calls user supplied functions
```

Comment:
Looks to me that it's just doing single loop.

Reply:
or do you mean it loops at `_batchRegister` and at `_batchPayBeneficiary` ? If so, worth mentioning in the comment